### DISCUSSION TITLE
how to remove sub menus Margin

### DISCUSSION THREAD

**Discussion by WillisXue** [AvaloniaUI/Avalonia#13600] [Q&A]
![image](https://github.com/AvaloniaUI/Avalonia/assets/9125768/1860ee97-6a22-4956-b67f-5bfe8285a03a)

I tried it but it not work

       <Style Selector="MenuItem /template/ ItemsPresenter">
           <Setter Property="Margin" Value="0"></Setter>
       </Style>

**Comment by timunie** [comment]
which theme is that? Simple? If so, check original Styles. I guess it may be this line: 
https://github.com/AvaloniaUI/Avalonia/blob/0567a6dbe7dc122ba90e019320708e0cbc51b635/src/Avalonia.Themes.Simple/Controls/FlyoutPresenter.xaml#L11

**Comment by maxkatz6** [comment]
https://github.com/AvaloniaUI/Avalonia/blob/eb7b6fe8435dfc3eae45907010b293f1ebc8d129/src/Avalonia.Themes.Simple/Controls/MenuItem.xaml#L12

**Reply by tms320** [reply]
I have the same problem. How to remove extra spaces on the left/right?
![image](https://github.com/user-attachments/assets/77cbec24-346c-42b8-a387-3873b75d60ec)

**Reply by tms320** [reply]
Data template didn't help too:
![image](https://github.com/user-attachments/assets/0f297a46-ef07-4623-9329-0864b6f4c159)
PS: I am using SimpleTheme

**Reply by timunie** [reply]
https://github.com/AvaloniaUI/Avalonia/blob/eb7b6fe8435dfc3eae45907010b293f1ebc8d129/src%2FAvalonia.Themes.Simple%2FControls%2FMenuItem.xaml#L74-L77

ItemsPresenter seems to have a Margin hardcoded, which means you need to copy the entire ControlTemplate unfortunately.

**Comment by tms320** [comment]
I created simple ControlTemplate based on default MenuItem template from 'Avalonia/src/Avalonia.Themes.Simple/Controls/MenuItem.xaml':

```xaml
<Window.Resources>
  <ControlTemplate x:Key="CustomMenuItemTemplate" TargetType="MenuItem">
	  <Border Name="root"
			  Background="{TemplateBinding Background}"
			  BorderBrush="{TemplateBinding BorderBrush}"
			  BorderThickness="{TemplateBinding BorderThickness}"
			  CornerRadius="{TemplateBinding CornerRadius}">
		  <Grid>
			  <Grid.ColumnDefinitions>
				  <ColumnDefinition Width="*" />
			  </Grid.ColumnDefinitions>
			  <ContentPresenter Name="PART_HeaderPresenter"
					Grid.Column="0"
					Margin="{TemplateBinding Padding}"
					VerticalAlignment="Center"
					Content="{TemplateBinding Header}"
					ContentTemplate="{TemplateBinding HeaderTemplate}">
				  <ContentPresenter.DataTemplates>
					  <DataTemplate DataType="sys:String">
						  <AccessText Text="{Binding}" />
					  </DataTemplate>
				  </ContentPresenter.DataTemplates>
			  </ContentPresenter>
		  </Grid>
	  </Border>
  </ControlTemplate>
</Window.Resources>
```
I can set the template for explicitly defined menu sub-items:
```xaml
<Menu>
	<MenuItem Header="Interfaces">
		<MenuItem Header="Interface 1" Template="{StaticResource CustomMenuItemTemplate}"/>
		<MenuItem Header="Interface 2" Template="{StaticResource CustomMenuItemTemplate}"/>
	</MenuItem>
</Menu>
```
But I don't understand how to set template if menu items are set via ItemsSource:
```xaml
<Menu>
	<MenuItem Header="Interfaces" ItemsSource="{Binding Interfaces}>
	</MenuItem>
</Menu>
```
I tried to set Template using style:
```xaml
<MenuItem Name="" Header="Interfaces" ItemsSource="{Binding Interfaces}">
	<MenuItem.Styles>
		<Style Selector="MenuItem">
			<Setter Property="Template" Value="{StaticResource CustomMenuItemTemplate}"/>
		</Style>
	</MenuItem.Styles>
</MenuItem>
```
But it didn't help.
How can I do that?

**Reply by timunie** [reply]
Maybe you need to set ContrainerTheme instead, which is a ControlTheme instead of a Style 🤔

**Reply by tms320** [reply]
The problem solved:
```xaml
<Menu>
	<MenuItem Header="Interfaces" ItemsSource="{Binding Interfaces}">
		<MenuItem.ItemContainerTheme>
			<ControlTheme TargetType="MenuItem">
				<Setter Property="Template" Value="{StaticResource CustomMenuItemTemplate}"/>
			</ControlTheme>
		</MenuItem.ItemContainerTheme>
	</MenuItem>
</Menu>
```
Thanks for help )
